### PR TITLE
Add DNS provider for Njalla

### LINKED ISSUE #1343: Add support for Njalla DNS provider
- Site : https://njal.la/
- API : https://njal.la/api/

### PR DESCRIPTION
**Pull request by ldez**
- [x] have a homogeneous design with the other providers
- [x] pass the linter
- [x] do `go mod tidy`
- [x] add tests (units)
- [x] add tests ("live")
- [x] add a [provider descriptor](https://github.com/go-acme/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [x] generate CLI help, documentation, and readme.
- [x] be able to do: _(and put the output of this command to a comment)_
```bash
rm -rf .lego

NJALLA_TOKEN=xxx ./lego -m your_email@example.com --dns njalla -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.

@dprandzioch could you test by running the previous command (you have to use a real domain and email)?

Closes #1343

### CODE REVIEW

File: providers/dns/njalla/internal/client.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+const apiEndpoint = "https://njal.la/api/1/"
+
+// Client is a Njallal API client.
```
Comment (Lctrs) [inline providers/dns/njalla/internal/client.go:13]:
```suggestion
// Client is a Njalla API client.
```

**Comment by Lctrs** [comment]
Thank you very much @ldez.

**Comment by dprandzioch** [comment]
For me this currently happens, I'm wondering why it's using namedotcom instead of njalla?

FYI I built the app from your feature branch in Docker to save setting up all the golang deps locally, hope this is fine.

```
seventeen:lego(feat/njalla) $ docker run -it --rm -e "NJALLA_TOKEN=*snap*" lego  -m letsencrypt@*snap* --dns njalla -d *.*snap* -d *snap* -s https://acme-staging-v02.api.letsencrypt.org/directory run
2021/02/15 20:26:52 No key found for account letsencrypt@*snap*. Generating a P256 key.
2021/02/15 20:26:52 Saved key to /.lego/accounts/acme-staging-v02.api.letsencrypt.org/letsencrypt@*snap*/keys/letsencrypt@*snap*.key
2021/02/15 20:26:53 namedotcom: some credentials information are missing: NAMECOM_USERNAME,NAMECOM_API_TOKEN
```

File: providers/dns/dns_providers.go [hunk 2]
```diff
@@ -212,6 +212,8 @@ func NewDNSChallengeProviderByName(name string) (challenge.Provider, error) {
 		return netlify.NewDNSProvider()
 	case "nifcloud":
 		return nifcloud.NewDNSProvider()
+	case "njalla":
+		return namedotcom.NewDNSProvider()
```
Comment (dprandzioch) [inline providers/dns/dns_providers.go:216]: Got it, this seems to be it. Did you forget to replace that?

Comment (ldez) [reply providers/dns/dns_providers.go]: Fixed

**Comment by dprandzioch** [comment]
OK got a step further but now I get errors finding the zone, which is strange because I can see it in the web interface. Do you have any idea how I can debug this further? Thanks for your work :-)

```
seventeen:lego(feat/njalla) $ docker run -it --rm -e "NJALLA_TOKEN=*snap*" lego  -m letsencrypt@*snap* --dns njalla -d *.*snap* -d *snap* -s https://acme-staging-v02.api.letsencrypt.org/directory run   
2021/02/15 21:00:02 No key found for account letsencrypt@*snap*. Generating a P256 key.
2021/02/15 21:00:02 Saved key to /.lego/accounts/acme-staging-v02.api.letsencrypt.org/letsencrypt@*snap*/keys/letsencrypt@*snap*.key
2021/02/15 21:00:03 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
Y
2021/02/15 21:00:07 [INFO] acme: Registering account for letsencrypt@*snap*
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2021/02/15 21:00:07 [INFO] [*.*snap*, *snap*] acme: Obtaining bundled SAN certificate
2021/02/15 21:00:08 [INFO] [*.*snap*] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/*snap*
2021/02/15 21:00:08 [INFO] [*snap*] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/*snap*
2021/02/15 21:00:08 [INFO] [*.*snap*] acme: use dns-01 solver
2021/02/15 21:00:08 [INFO] [*snap*] acme: Could not find solver for: tls-alpn-01
2021/02/15 21:00:08 [INFO] [*snap*] acme: Could not find solver for: http-01
2021/02/15 21:00:08 [INFO] [*snap*] acme: use dns-01 solver
2021/02/15 21:00:08 [INFO] [*.*snap*] acme: Preparing to solve DNS-01
2021/02/15 21:00:08 [INFO] [*snap*] acme: Preparing to solve DNS-01
2021/02/15 21:00:08 [INFO] [*.*snap*] acme: Cleaning DNS-01 challenge
2021/02/15 21:00:08 [WARN] [*.*snap*] acme: cleaning up failed: njalla: unknown record ID for '_acme-challenge.*snap*.' '*snap*' 
2021/02/15 21:00:08 [INFO] [*snap*] acme: Cleaning DNS-01 challenge
2021/02/15 21:00:08 [WARN] [*snap*] acme: cleaning up failed: njalla: unknown record ID for '_acme-challenge.*snap*.' '*snap*' 
2021/02/15 21:00:09 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/*snap*
2021/02/15 21:00:09 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/*snap*
2021/02/15 21:00:09 Could not obtain certificates:
	error: one or more domains had a problem:
[*.*snap*] [*.*snap*] acme: error presenting token: njalla: failed to find zone: code: 403, message: permission denied
[*snap*] [*snap*] acme: error presenting token: njalla: failed to find zone: code: 403, message: permission denied
```

**Comment by ldez** [comment]
could you replace your domain, in your comment by `example.com` instead of the weird `*snap*`

**Comment by ldez** [comment]
> njalla: failed to find zone: code: 403, message: permission denied

it's not the right message (my mistake), the real message will be `njalla: failed to add record, code: 403, message: permission denied`

**Comment by ldez** [comment]
I think there is an issue with your token.

Why didn't you lego inside a container?

**Comment by ldez** [comment]
When I use an invalid  token I have: `code: 403, message: Invalid token.`

> code: 403, message: permission denied

based on your message I think that the token seems valid but it doesn't provide the rights to edit the domain records.

Could you check your token?

**Comment by dprandzioch** [comment]
@ldez Then I have to set up all the golang deps on the host first. Pretty sure it's not about that but i'll verify. I'll let you know once I have further results

**Comment by dprandzioch** [comment]
OK when I run a example cURL using the same token and domain, it works:

```
seventeen:lego(feat/njalla) $ curl -H "Authorization: Njalla abc" -H "Content-Type: application/json" --data '{"method":"get-domain", "params": {"domain": "example.com"}}' https://njal.la/api/1/
{"result": {"name": "example.com", "status": "active", "expiry": "2022-02-18T12:40:17Z", "locked": true, "mailforwarding": false, "max_nameservers": 10, "dnssec_type": "dsData"}, "jsonrpc": "2.0"}
```

**Comment by dprandzioch** [comment]
@ldez This happens when running on the host with the tested token from above

```
seventeen:dist(feat/njalla) $ NJALLA_TOKEN=foo ./lego -m letsencrypt@example.com --dns njalla -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
2021/02/16 19:32:02 No key found for account letsencrypt@example.com. Generating a P256 key.
2021/02/16 19:32:02 Saved key to /Users/davd/Workspace/lego/dist/.lego/accounts/acme-staging-v02.api.letsencrypt.org/letsencrypt@example.com/keys/letsencrypt@example.com.key
2021/02/16 19:32:03 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
Y
2021/02/16 19:32:05 [INFO] acme: Registering account for letsencrypt@example.com
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/Users/davd/Workspace/lego/dist/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2021/02/16 19:32:05 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/02/16 19:32:06 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/foo
2021/02/16 19:32:06 [INFO] [example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/foo
2021/02/16 19:32:06 [INFO] [*.example.com] acme: use dns-01 solver
2021/02/16 19:32:06 [INFO] [example.com] acme: Could not find solver for: tls-alpn-01
2021/02/16 19:32:06 [INFO] [example.com] acme: Could not find solver for: http-01
2021/02/16 19:32:06 [INFO] [example.com] acme: use dns-01 solver
2021/02/16 19:32:06 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/02/16 19:32:07 [INFO] [example.com] acme: Preparing to solve DNS-01
2021/02/16 19:32:07 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/02/16 19:32:07 [WARN] [*.example.com] acme: cleaning up failed: njalla: unknown record ID for '_acme-challenge.example.com.' 'foo' 
2021/02/16 19:32:07 [INFO] [example.com] acme: Cleaning DNS-01 challenge
2021/02/16 19:32:07 [WARN] [example.com] acme: cleaning up failed: njalla: unknown record ID for '_acme-challenge.example.com.' 'foo' 
2021/02/16 19:32:07 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/foo
2021/02/16 19:32:07 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/foo
2021/02/16 19:32:07 Could not obtain certificates:
	error: one or more domains had a problem:
[*.example.com] [*.example.com] acme: error presenting token: njalla: failed to add record: code: 403, message: permission denied
[example.com] [example.com] acme: error presenting token: njalla: failed to add record: code: 403, message: permission denied
```

**Comment by ldez** [comment]
@dprandzioch I added some debug logs.

**Comment by dprandzioch** [comment]
Hey, sorry for the delay. Here's what happens. Unfortunately I cannot read much from that, maybe you can?

```
seventeen:dist(feat/njalla) $ NJALLA_TOKEN=foo ./lego -m letsencrypt@example.com --dns njalla -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
2021/02/19 19:56:59 No key found for account letsencrypt@example.com. Generating a P256 key.
2021/02/19 19:56:59 Saved key to /Users/davd/Workspace/lego/dist/.lego/accounts/acme-staging-v02.api.letsencrypt.org/letsencrypt@example.com/keys/letsencrypt@example.com.key
2021/02/19 19:57:00 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
Y
2021/02/19 19:57:02 [INFO] acme: Registering account for letsencrypt@example.com
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/Users/davd/Workspace/lego/dist/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2021/02/19 19:57:02 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/02/19 19:57:03 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/123
2021/02/19 19:57:03 [INFO] [example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/1234
2021/02/19 19:57:03 [INFO] [*.example.com] acme: use dns-01 solver
2021/02/19 19:57:03 [INFO] [example.com] acme: Could not find solver for: tls-alpn-01
2021/02/19 19:57:03 [INFO] [example.com] acme: Could not find solver for: http-01
2021/02/19 19:57:03 [INFO] [example.com] acme: use dns-01 solver
2021/02/19 19:57:03 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/02/19 19:57:03 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla foo] Content-Type:[application/json]] {0x140006a2b10} 0x102c62d10 158 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x140000440e8} {"method":"add-record","params":{"content":"abc","domain":"example.com.","name":"_acme-challenge","ttl":300,"type":"TXT"}}
2021/02/19 19:57:04 [INFO] [example.com] acme: Preparing to solve DNS-01
2021/02/19 19:57:04 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla foo] Content-Type:[application/json]] {0x140006a3470} 0x102c62d10 158 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x140000440e8} {"method":"add-record","params":{"content":"abcd","domain":"example.com.","name":"_acme-challenge","ttl":300,"type":"TXT"}}
2021/02/19 19:57:04 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/02/19 19:57:04 [WARN] [*.example.com] acme: cleaning up failed: njalla: unknown record ID for '_acme-challenge.example.com.' 'def'
2021/02/19 19:57:04 [INFO] [example.com] acme: Cleaning DNS-01 challenge
2021/02/19 19:57:04 [WARN] [example.com] acme: cleaning up failed: njalla: unknown record ID for '_acme-challenge.example.com.' 'defg'
2021/02/19 19:57:04 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/123
2021/02/19 19:57:05 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/1234
2021/02/19 19:57:05 Could not obtain certificates:
	error: one or more domains had a problem:
[*.example.com] [*.example.com] acme: error presenting token: njalla: failed to add record: code: 403, message: permission denied
[example.com] [example.com] acme: error presenting token: njalla: failed to add record: code: 403, message: permission denied
```

**Comment by ldez** [comment]
@dprandzioch I think I found the issue.

could you try the following commands (replace `example.com` and `abc` by your domain information):

```bash
curl \
  -H "Authorization: Njalla abc" \
  -H "Content-Type: application/json" \
  -H "Accept: application/json" \
  --data '{"method":"add-record","params":{"content":"abc","domain":"example.com.","name":"_acme-challenge","ttl":300,"type":"TXT"}}' \
  https://njal.la/api/1/
```

```
curl \
  -H "Authorization: Njalla abc" \
  -H "Content-Type: application/json" \
  -H "Accept: application/json" \
  --data '{"method":"add-record","params":{"content":"abc","domain":"example.com","name":"_acme-challenge","ttl":300,"type":"TXT"}}' \
  https://njal.la/api/1/
```

**Comment by dprandzioch** [comment]
Yeah you're right, removing the fqdn dot at the end fixes it:

```
seventeen:dist(feat/njalla) $ curl \
  -H "Authorization: Njalla abc" \
  -H "Content-Type: application/json" \
  -H "Accept: application/json" \
  --data '{"method":"add-record","params":{"content":"abc","domain":"example.com.","name":"_acme-challenge","ttl":300,"type":"TXT"}}' \
  https://njal.la/api/1/
{"error": {"code": 403, "message": "permission denied"}, "jsonrpc": "2.0"}

seventeen:dist(feat/njalla) $ curl \
  -H "Authorization: Njalla abc" \
  -H "Content-Type: application/json" \
  -H "Accept: application/json" \
  --data '{"method":"add-record","params":{"content":"abc","domain":"example.com","name":"_acme-challenge","ttl":300,"type":"TXT"}}' \
  https://njal.la/api/1/
{"result": {"id": 428477, "name": "_acme-challenge", "type": "TXT", "content": "abc", "ttl": 300}, "jsonrpc": "2.0"}
```

**Comment by ldez** [comment]
could you retry the PR now?

**Comment by dprandzioch** [comment]
Got further now:

```
seventeen:dist(feat/njalla) $ NJALLA_TOKEN=tokenabc ./lego -m letsencrypt@example.com --dns njalla -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
2021/02/19 20:19:39 No key found for account letsencrypt@example.com. Generating a P256 key.
2021/02/19 20:19:39 Saved key to /Users/davd/Workspace/lego/dist/.lego/accounts/acme-staging-v02.api.letsencrypt.org/letsencrypt@example.com/keys/letsencrypt@example.com.key
2021/02/19 20:19:40 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
Do you accept the TOS? Y/n
Y
2021/02/19 20:19:42 [INFO] acme: Registering account for letsencrypt@example.com
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/Users/davd/Workspace/lego/dist/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2021/02/19 20:19:42 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/02/19 20:19:43 [INFO] [*.example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/123
2021/02/19 20:19:43 [INFO] [example.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/234
2021/02/19 20:19:43 [INFO] [*.example.com] acme: use dns-01 solver
2021/02/19 20:19:43 [INFO] [example.com] acme: Could not find solver for: tls-alpn-01
2021/02/19 20:19:43 [INFO] [example.com] acme: Could not find solver for: http-01
2021/02/19 20:19:43 [INFO] [example.com] acme: use dns-01 solver
2021/02/19 20:19:43 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/02/19 20:19:43 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla tokenabc] Content-Type:[application/json]] {0x1400030f4d0} 0x104332d10 157 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400012e008} {"method":"add-record","params":{"content":"tokena","domain":"example.com","name":"_acme-challenge","ttl":300,"type":"TXT"}}
2021/02/19 20:19:43 [INFO] [example.com] acme: Preparing to solve DNS-01
2021/02/19 20:19:43 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla tokenabc] Content-Type:[application/json]] {0x14000099170} 0x104332d10 157 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400012e008} {"method":"add-record","params":{"content":"tokenb","domain":"example.com","name":"_acme-challenge","ttl":300,"type":"TXT"}}
2021/02/19 20:19:44 [INFO] [*.example.com] acme: Trying to solve DNS-01
2021/02/19 20:19:44 [INFO] [*.example.com] acme: Checking DNS record propagation using [192.168.1.3:53]
2021/02/19 20:19:46 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/02/19 20:19:53 [INFO] [example.com] acme: Trying to solve DNS-01
2021/02/19 20:19:53 [INFO] [example.com] acme: Checking DNS record propagation using [192.168.1.3:53]
2021/02/19 20:19:55 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/02/19 20:20:03 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/02/19 20:20:03 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla tokenabc] Content-Type:[application/json]] {0x140000994a0} 0x104332d10 68 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400012e008} {"method":"remove-record","params":{"id":428478,"domain":"example.com"}}
2021/02/19 20:20:03 [INFO] [example.com] acme: Cleaning DNS-01 challenge
2021/02/19 20:20:03 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla tokenabc] Content-Type:[application/json]] {0x140001d19e0} 0x104332d10 68 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400012e008} {"method":"remove-record","params":{"id":428479,"domain":"example.com"}}
2021/02/19 20:20:03 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/123
2021/02/19 20:20:03 [INFO] Unable to deactivate the authorization: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/123
2021/02/19 20:20:04 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/234
2021/02/19 20:20:04 [INFO] Unable to deactivate the authorization: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/234
2021/02/19 20:20:04 Could not obtain certificates:
	error: one or more domains had a problem:
[*.example.com] acme: error: 400 :: urn:ietf:params:acme:error:dns :: DNS problem: SERVFAIL looking up TXT for _acme-challenge.example.com - the domain's nameservers may be malfunctioning
[example.com] acme: error: 400 :: urn:ietf:params:acme:error:dns :: DNS problem: SERVFAIL looking up TXT for _acme-challenge.example.com - the domain's nameservers may be malfunctioning
```

I'm just wondering why the cert creation does not work. Probably it's my pihole not refreshing fast enough, I'll retry with googles public dns in a sec.

**Comment by dprandzioch** [comment]
Yeah, so I changed to cloudflares upstream dns but I had some problems with the LE staging, so I just tried on production and get certs served:

```
seventeen:dist(feat/njalla) $ NJALLA_TOKEN=foo ./lego -m letsencrypt@example.com --dns njalla -d *.example.com -d example.com run
2021/02/19 20:24:40 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2021/02/19 20:24:41 [INFO] [*.example.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/123
2021/02/19 20:24:41 [INFO] [example.com] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/456
2021/02/19 20:24:41 [INFO] [*.example.com] acme: use dns-01 solver
2021/02/19 20:24:41 [INFO] [example.com] acme: Could not find solver for: tls-alpn-01
2021/02/19 20:24:41 [INFO] [example.com] acme: Could not find solver for: http-01
2021/02/19 20:24:41 [INFO] [example.com] acme: use dns-01 solver
2021/02/19 20:24:41 [INFO] [*.example.com] acme: Preparing to solve DNS-01
2021/02/19 20:24:41 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla foo] Content-Type:[application/json]] {0x140001df620} 0x10252ad10 162 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400021a000} {"method":"add-record","params":{"content":"111","domain":"example.com","name":"_acme-challenge","ttl":300,"type":"TXT"}}
2021/02/19 20:24:41 [INFO] [example.com] acme: Preparing to solve DNS-01
2021/02/19 20:24:41 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla foo] Content-Type:[application/json]] {0x140001ac390} 0x10252ad10 162 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400021a000} {"method":"add-record","params":{"content":"222","domain":"example.com","name":"_acme-challenge","ttl":300,"type":"TXT"}}
2021/02/19 20:24:41 [INFO] [*.example.com] acme: Trying to solve DNS-01
2021/02/19 20:24:41 [INFO] [*.example.com] acme: Checking DNS record propagation using [1.1.1.1:53]
2021/02/19 20:24:43 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/02/19 20:24:43 [INFO] [*.example.com] acme: Waiting for DNS record propagation.
2021/02/19 20:24:51 [INFO] [*.example.com] The server validated our request
2021/02/19 20:24:51 [INFO] [example.com] acme: Trying to solve DNS-01
2021/02/19 20:24:51 [INFO] [example.com] acme: Checking DNS record propagation using [1.1.1.1:53]
2021/02/19 20:24:53 [INFO] Wait for propagation [timeout: 1m0s, interval: 2s]
2021/02/19 20:25:07 [INFO] [example.com] The server validated our request
2021/02/19 20:25:07 [INFO] [*.example.com] acme: Cleaning DNS-01 challenge
2021/02/19 20:25:07 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla foo] Content-Type:[application/json]] {0x140001905a0} 0x10252ad10 68 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400021a000} {"method":"remove-record","params":{"id":428482,"domain":"example.com"}}
2021/02/19 20:25:07 [INFO] [example.com] acme: Cleaning DNS-01 challenge
2021/02/19 20:25:07 &{POST https://njal.la/api/1/ HTTP/1.1 1 1 map[Accept:[application/json] Authorization:[Njalla foo] Content-Type:[application/json]] {0x140001ac210} 0x10252ad10 68 [] false njal.la map[] map[] <nil> map[]   <nil> <nil> <nil> 0x1400021a000} {"method":"remove-record","params":{"id":428483,"domain":"example.com"}}
2021/02/19 20:25:07 [INFO] [*.example.com, example.com] acme: Validations succeeded; requesting certificates
2021/02/19 20:25:11 [INFO] [*.example.com] Server responded with a certificate.
```

So I guess it works

**Comment by ldez** [comment]
Thank you @dprandzioch

**Comment by dprandzioch** [comment]
@ldez Thanks for the effort you've taken. Will be a happy user serving valid certificates on my domain for my home server :-)

**Review by dmke — APPROVED** [review]

**Comment by dprandzioch** [comment]
🥳

**Pull request merged**